### PR TITLE
Fix behavioural differences due to return type and handle constant string literal inputs differently for COALESCE function

### PR DESCRIPTION
**Pull request by basasairohan**
### Description

Behavioural differences mentioned below have been found out better PG Coalesce and T-SQL Coalesce function.
1. Consistency in return datatype i.e., for the same set of input datatypes T-SQL should return consistent return datatype as shown in below example
```
-- T-SQL
select coalesce(cast(8 as smallint), cast(1 as bit))
go
~~START~~
smallint
8
~~END~~
select coalesce(cast(1 as bit), cast(8 as smallint))
go
~~START~~
smallint
1
~~END~~

-- BBF

select coalesce(cast(8 as smallint), cast(1 as bit))
go
~~START~~
smallint
8
~~END~~
select coalesce(cast(1 as bit), cast(8 as smallint))
go
~~START~~
bit
1
~~END~~
```
2. When the input is constant string literal i.e., T-SQL should take short circuit evaluation meaning unnecessary arguments should not be evaluated
```
-- T-SQL
1> SELECT COALESCE(NULL, 1, 2, 'I am a string')
2> go
           
-----------
          1

(1 rows affected)

-- BBF
1> SELECT COALESCE(NULL, 1, 2, 'I am a string')
2> go
Msg 22, Level 16, State 1, Server BABEL, Line 1
invalid input syntax for integer: "I am a string"
```

This change adds a new function `select_common_type_for_coalesce_function` to choose the return datatype which will be called from `tsql_select_common_type_hook` whenever the caller function is T-SQL COALESCE.

### Issues Resolved

[BABEL-726]

Engine PR : https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/345

Signed-off-by: Sai Rohan Basa [bsrohan@amazon.com](mailto:bsrohan@amazon.com)

### Test Scenarios Covered ###
* **Use case based -** babel_726.sql

* **Boundary conditions -** Yes

* **Arbitrary inputs -** Yes

* **Negative test cases -** Yes

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -** N/A

* **Tooling impact -** N/A

* **Client tests -** N/A

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 1]
```diff
@@ -1267,6 +1269,8 @@ tsql_select_common_type_hook(ParseState *pstate, List *exprs, const char *contex
 		return InvalidOid;
 	else if (strncmp(context, "ISNULL", strlen("ISNULL")) == 0)
 		return select_common_type_for_isnull(pstate, exprs);
+	else if(strncmp(context, "TSQL_COALESCE", strlen("TSQL_COALESCE")) == 0)
+		return select_common_type_for_coalesce_function(pstate,exprs);
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: nit: space between `,` and `exprs`

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 2]
```diff
@@ -1357,6 +1361,60 @@ select_common_type_for_isnull(ParseState *pstate, List *exprs)
...
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("COALESCE function should have at least 2 arguments")));
+
+	foreach(lc, exprs)
+	{
+		pexpr = (Node *) lfirst(lc);
+		ptype = exprType(pexpr);
+
+		/* Check if arg is NULL literal */
+		if (IsA(pexpr, Const) && ((Const *) pexpr)->constisnull && ptype == UNKNOWNOID)
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: What is the need to check `ptype == UNKNOWNOID` here?

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 3]
```diff
@@ -1357,6 +1361,60 @@ select_common_type_for_isnull(ParseState *pstate, List *exprs)
...
+			continue;
+
+		/* If the arg is non-null string literal */
+		if (ptype == UNKNOWNOID)
+		{
+			Oid curr_oid = get_sys_varcharoid();
+			if (commontype == InvalidOid 
+				|| tsql_get_type_precedence(curr_oid) < curr_precedence)
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: curr_precedence is uninitialized

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 4]
```diff
@@ -1357,6 +1361,60 @@ select_common_type_for_isnull(ParseState *pstate, List *exprs)
...
+			{
+				commontype = curr_oid;
+				curr_precedence = tsql_get_type_precedence(curr_oid);
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Define `tsql_get_type_precedence(curr_oid)` and `tsql_get_type_precedence(ptype)` once and then reuse their values

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 2 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Also, the `if ... continue, if ... continue, if` could be replaced by `if ... else if ... else if`

File: test/JDBC/expected/babel_726.out [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+
+
+select coalesce(NULL)
+go
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: COALESCE function should have at least 2 arguments)~~
+
+
+-- Empty or white spaced strings
```
Comment (thephantomthief) [inline test/JDBC/expected/babel_726.out:16]: Also try with `\n`, `\t` characters

Comment (basasairohan) [reply test/JDBC/expected/babel_726.out:16]: done, added tests

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 4 above)
Comment (basasairohan) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: done

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 3 above)
Comment (basasairohan) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: done

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 6]
```diff
@@ -1277,6 +1282,45 @@ tsql_select_common_type_hook(ParseState *pstate, List *exprs, const char *contex
...
+
+	con = (Const *) e;
+	val = DatumGetCString(con->constvalue);
+
+	if (val != NULL)
+		i = strlen(val) - 1;
+
+	/*
+	 *	Additional handling for empty or white space string literals as
+	 *	T-SQL treats an empty string literal as 0 in certain datatypes
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Nit: Indentation. There seems to be an extra tab after `*`

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 7]
```diff
@@ -1357,6 +1401,63 @@ select_common_type_for_isnull(ParseState *pstate, List *exprs)
...
+		{
+			commontype = ptype;
+			curr_precedence = temp_precedence;
+		}
+	}
+
+	if (commontype == InvalidOid)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("At least one of the arguments to COALESCE must be a non NULL constant")));
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Nit: non-NULL

File: test/JDBC/expected/babel_726.out [hunk 8]
```diff
@@ -0,0 +1,591 @@
...
+
+~~ERROR (Message: invalid input syntax for type integer: "I am a string")~~
+
+
+-- precedence correctness
+EXEC sp_babelfish_configure 'babelfishpg_tsql.escape_hatch_rowversion', 'ignore';
+go
+
+create table babel_726_t1 (
+    a1 sql_variant,
```
Comment (thephantomthief) [inline test/JDBC/expected/babel_726.out:72]: Please add numeric as well

File: test/JDBC/expected/babel_726.out [hunk 9]
```diff
@@ -0,0 +1,591 @@
...
+int
+2
+~~END~~
+
+
+-- tab space
+select coalesce(NULL, CHAR(9))
+go
+~~START~~
+text
```
Comment (thephantomthief) [inline test/JDBC/expected/babel_726.out:37]: is it expected we get text type as result?

**Review by thephantomthief — APPROVED** [review]

**Review by jsudrik — APPROVED** [review]
Approved based on the other reviews.

**Pull request merged**